lib/beaker/subcommand.rb

```diff
@@ -49,55 +53,134 @@ def initialize(*args)
-      say "Writing host config to .beaker/acceptance/config/default_#{hypervisor}_hosts.yaml"
-      SubcommandUtil.store_config({:hypervisor => hypervisor})
-      SubcommandUtil.delete_config([:provisioned])
+
+      @cli.parse_options
+
+      # delete unnecessary keys for saving the options
+      options_to_write = @cli.configured_options
+      # Remove keys we don't want to save
+      [:timestamp, :logger, :command_line, :beaker_version, :hosts_file].each do |key|
+        options_to_write.delete(key)
+      end
+
+      options_to_write = SubcommandUtil.sanitize_options_for_save(options_to_write)
+
+      @cli.logger.notify 'writing configured options to disk'
+      File.open(SubcommandUtil::SUBCOMMAND_OPTIONS, 'w') do |f|
+        f.write(options_to_write.to_yaml)
+      end
+
+      state = YAML::Store.new(SubcommandUtil::SUBCOMMAND_STATE)
+      state.transaction do
+        state['provisioned'] = false
+      end
     end
 
-    desc "provision", "Provisions the beaker test configuration"
-    option :validate, :type => :boolean, :default => true
-    option :configure, :type => :boolean, :default => true
+    desc "provision", "Provisions the beaker systems under test(SUTs)"
     option :help, :type => :boolean, :hide => true
     long_desc <<-LONGDESC
-    Provisions a beaker configuration
+    Provisions hosts defined in your subcommand_options file. You can pass the --hosts
+    flag here to override any hosts provided there. Really, you can pass most any beaker
+    flag here to override.
     LONGDESC
     def provision()
       if options[:help]
         invoke :help, [], ["provision"]
         return
       end
 
-      hypervisor = @@config.transaction { @@config[:hypervisor] }
+      state = YAML::Store.new(SubcommandUtil::SUBCOMMAND_STATE)
+      if state.transaction { state['provisioned']}
+        SubcommandUtil.error_with('Provisioned SUTs detected. Please destroy and reprovision.')
+      end
+      @cli.parse_options
+      @cli.provision
+
+      # Sanitize the hosts
+      cleaned_hosts = SubcommandUtil.sanitize_options_for_save(@cli.combined_instance_and_options_hosts)
 
-      unless hypervisor
-        SubcommandUtil.error_with("Please initialise a configuration")
+      # Update each host provisioned with a flag indicating that it no longer needs
+      # provisioning
+      cleaned_hosts.each do |host, host_hash|
+        host_hash['provision'] = false
       end
 
-      provisioned = @@config.transaction { @@config[:provisioned] }
 
-      if !provisioned or options[:force]
-        SubcommandUtil.provision(hypervisor, options)
-        SubcommandUtil.store_config({:provisioned => true})
+      # should we only update the options here with the new host? Or update the settings
+      # with whatever new flags may have been provided with provision?
+      options_storage = YAML::Store.new(SubcommandUtil::SUBCOMMAND_OPTIONS)
+      options_storage.transaction do
```

Comment:
I'm also into this idea, but I think we can revisit after we have established our pattern of exactly what we need from the accessor/getter function.